app/forms/webauthn_verification_form.rb

```diff
@@ -53,23 +62,41 @@ def self.domain_name
 
   private
 
-  attr_reader :user,
+  attr_reader :protocol,
+              :user,
               :challenge,
-              :protocol,
               :authenticator_data,
               :client_data_json,
               :signature,
               :credential_id,
-              :webauthn_error
+              :webauthn_error,
+              :screen_lock_error
+
+  def validate_screen_lock_error
+    return if screen_lock_error.blank?
+    errors.add(:screen_lock_error, screen_lock_error_message, type: :screen_lock_error)
+  end
+
+  def validate_required_attributes_present
+    [
+      :challenge,
+      :authenticator_data,
+      :client_data_json,
+      :signature,
+      :webauthn_configuration,
+    ].each do |attr|
+      errors.add(attr, generic_error_message, type: :blank) if self.send(attr).blank?
+    end
+  end
```

Comment:
I saw there was some chatter in Slack about this, I'd do:
```suggestion
  validates :challenge, :authenticator_data, :client_data_json, :signature, :webauthn_configuration,
    presence: { message: ->(_object, _data) { generic_error_message } }
```

Reply:
I don't think that'll work since ([as mentioned in Slack](https://gsa-tts.slack.com/archives/C0NGESUN5/p1699453659556769?thread_ts=1699453536.012369&cid=C0NGESUN5)) the `message` `proc` is run in the context of the class, not the instance, so `generic_error_message`'s references to `platform_authenticator?` will be unavailable.

Reply:
That's good to know we can validate multiple presence at the same time like that though, I wasn't aware.

Reply:
In some cases we can probably derive the equivalent of certain instance variables from the `_data` second argument, but I don't know that that'd work here, and it feels a little hacky.

Reply:
in that case `object` is probably the instance so we can use that and do `object.generic_error_message`

Reply:
> in that case `object` is probably the instance so we can use that and do `object.generic_error_message`

Hm, I'll double-check that, since that could work yeah. When I was researching this earlier I had the impression that the first parameter was the `error`, not the `object`.

Reply:
That works a treat 👍 See 2b7138c07.

Only downside is moving `generic_error_message` from being a private method to public, since we're calling the method on the object "externally".